rest_sensors.cpp

```diff
@@ -2111,6 +2111,89 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
                     }
                 }
             }
+            
+            //Special part for metering interfaces
+            if (sensor->type() == "ZHAConsumption")
+            {
+                if (rid.suffix == RConfigPulseConfiguration)
+                {
+                    if (map[pi.key()].type() == QVariant::Double)
+                    {
+                        if (sensor->modelId() == QLatin1String("ZHEMI101"))
+                        {
+                            quint16 pulseConfiguration = map[pi.key()].toUInt(&ok);
+
+                            if (ok && addTaskSimpleMeteringReadWriteAttribute(task, deCONZ::ZclWriteAttributesId, 0x0300, deCONZ::Zcl16BitUint, pulseConfiguration, VENDOR_DEVELCO))
+                            {
+                                updated = true;
+                            }
+                            else
+                            {
+                                rsp.list.append(errorToMap(ERR_ACTION_ERROR, QString("/sensors/%1/config/%2").arg(id).arg(pi.key()).toHtmlEscaped(),
+                                                           QString("Could not set attribute")));
+                                rsp.httpStatus = HttpStatusBadRequest;
+                                return REQ_READY_SEND;
+                            }
+                        }
+                    }
```

Comment:
`if (sensor->modelId() == QLatin1String("ZHEMI101")) ...`

When this is called for sensors which have a `RConfigPulseConfiguration` but aren't "ZHEMI101" there seems to be a missing error else branch?

`quint16 pulseConfiguration = map[pi.key()].toUInt(&ok);`

Implicit cast, maybe not realistic but a client could call this with values > UINT16_MAX which would be silently truncated.

```
const uint pulseConfiguration = map[pi.key()].toUInt(&ok);

if (ok && pulseConfiguration <=  UINT16_MAX && addTaskSimpleMeteringReadWriteAttribute(task, deCONZ::ZclWriteAttributesId, 0x0300, deCONZ::Zcl16BitUint, pulseConfiguration, VENDOR_DEVELCO)
{
}
```

Reply:
> When this is called for sensors which have a RConfigPulseConfiguration but aren't "ZHEMI101" there seems to be a missing error else branch?

Well, this is true, however, that is elsewhere in the function also not the case. I assume it should throw a `ERR_PARAMETER_NOT_AVAILABLE`?

I'm currently really thinking this function could use a decent refactor to take out complexity and repetitions while making it more robust and shrinking size...

> Implicit cast, maybe not realistic but a client could call this with values > UINT16_MAX which would be silently truncated.

Totally agree

Reply:
>Well, this is true, however, that is elsewhere in the function also not the case. 

I guess we have unfortunately many places with missing error handling.

>I assume it should throw a ERR_PARAMETER_NOT_AVAILABLE?

Sounds fine to me.

>I'm currently really thinking this function could use a decent refactor to take out complexity and repetitions while making it more robust and shrinking size...

All in :D I'd really like to split this up into multiple smaller functions each only doing once thing/topic so we get a chance to get closer to automatic testing.